test/logflare/users/users_test.exs

```diff
@@ -45,4 +43,73 @@ defmodule Logflare.UsersTest do
       assert get_by(api_key: u1.api_key).email_preferred == nil
     end
   end
+
+  describe "insert_user/1" do
+    test "inserts new user" do
+      params = %{
+        "email" => Faker.Internet.email(),
+        "provider_uid" => Ecto.UUID.generate(),
+        "provider" => "email",
+        "token" => Ecto.UUID.generate()
+      }
+
+      assert {:ok, _user} = Users.insert_user(params)
+    end
+  end
+
+  describe "insert_or_update_user/1" do
+    test "if user exists with provider_uid, updates it", %{user: u1} do
+      name = Faker.Person.name()
+
+      params = %{
+        name: name,
+        provider: "email",
+        email: u1.email,
+        provider_uid: u1.provider_uid,
+        token: u1.token
+      }
+
+      {:ok_found_user, user} = Users.insert_or_update_user(params)
+      assert user.name == name
+      assert user.id == u1.id
+    end
+
+    test "if provider_uid does not exist but user exists, search by email and update", %{user: u1} do
+      name = Faker.Person.name()
+
+      params = %{
+        name: name,
+        provider: "email",
+        email: u1.email,
+        provider_uid: Ecto.UUID.generate(),
+        token: u1.token
+      }
+
+      {:ok_found_user, user} = Users.insert_or_update_user(params)
+      assert user.name == name
+      assert user.id == u1.id
+    end
+
+    test "if no user exists with given email or provider_uid creates a new one with given params" do
+      params = %{
```

Comment:
> You can set a @valid_params then use Map.take/2 to pick out what you need (or don't need using Map.drop/2, would lead to less duplication. 

Can you clarify this one?

Reply:
honestly I rather have a more explicit set of params than this. might be more repetition but ends up being way more explicit on what we're testing 🤔